std: uefi: Add basic Env variables

Ayush1325 opened the pull request:
- Implement environment variable functions
- Using EFI Shell protocol.

jhpratt commented:
Not familiar with UEFI

r? libs

tgross35 commented:
@joboet does I think, at least they reviewed https://github.com/rust-lang/rust/pull/123196

joboet commented:
I guess? I worked with UEFI before, so feel free to
r? @joboet 
in the future

joboet commented on library/std/src/sys/pal/uefi/os.rs:
```diff
@@ -239,3 +239,104 @@ pub fn exit(code: i32) -> ! {
+    pub(crate) fn get(key: &OsStr) -> Option<OsString> {
+        let rt: NonNull<r_efi::efi::RuntimeServices> =
+            helpers::runtime_services().expect("UEFI Runtime Services Missing").cast();
+        let mut key = key.encode_wide().chain(Some(0)).collect::<Vec<u16>>();
```
This should check for internal NULL bytes and then return `None`.

joboet replied:
This still needs to be addressed.

joboet commented on library/std/src/sys/pal/uefi/os.rs:
```diff
@@ -239,3 +239,104 @@ pub fn exit(code: i32) -> ! {
+    }
+
+    pub(crate) fn unset(key: &OsStr) -> io::Result<()> {
+        let mut key = key.encode_wide().chain(Some(0)).collect::<Vec<u16>>();
```
Here as well.

joboet commented on library/std/src/sys/pal/uefi/os.rs:
```diff
@@ -239,3 +239,104 @@ pub fn exit(code: i32) -> ! {
+    pub(crate) fn set(key: &OsStr, val: &OsStr) -> io::Result<()> {
+        let mut key = key.encode_wide().chain(Some(0)).collect::<Vec<u16>>();
+        // UEFI variable value does not need to be NULL terminated.
+        let mut val = val.encode_wide().collect::<Vec<u16>>();
```
It does, since we need to cooperate with the UEFI shell protocol (we use the same GUID after all).

Ayush1325 replied:
I think I wasn't clear with the comment wording. Edk2 does not write the NULL terminator to the variable value: https://github.com/tianocore/edk2/blob/049e12c03d27f0a6bf57f4f1835cab5e205661a7/ShellPkg/Application/Shell/ShellProtocol.c#L2956
In fact it will be inconsistent if we do write it.

joboet commented on library/std/src/sys/pal/uefi/os.rs:
```diff
@@ -239,3 +239,104 @@ pub fn exit(code: i32) -> ! {
+        if ret.is_error() {
+            None
+        } else {
+            unsafe { val.set_len(len / size_of::<u16>()) };
+            Some(OsString::from_wide(&val))
```
This should remove trailing NULLs.

Ayush1325 replied:
That was my initial assumption but see https://github.com/tianocore/edk2/blob/049e12c03d27f0a6bf57f4f1835cab5e205661a7/ShellPkg/Application/Shell/ShellProtocol.c#L2956
Edk2 does not write null for value.

joboet replied:
Ah, right. Makes sense!

joboet reviewed (commented):
I don't remember: was there a decision to always use the underlying UEFI calls instead of UEFI-shell calls? The current implementation will interact nicely with the EDK2 shell implementation, but won't work the same way with other UEFI implementations. But we don't have to decide here.

Could you also implement `Env` using `GetNextVariableName`? I think we should support that as well.

Ayush1325 commented:
> I don't remember: was there a decision to always use the underlying UEFI calls instead of UEFI-shell calls? The current implementation will interact nicely with the EDK2 shell implementation, but won't work the same way with other UEFI implementations. But we don't have to decide here.

So, the reason for not using shell protocol till now was that it will not be present most of the time. So as long as we do not need to depend on it, we should not.

With that said, as long as we can provide a fallback to default UEFI variables, I think it is fine to depend on the Shell Protocol for environment variables.

> 
> Could you also implement `Env` using `GetNextVariableName`? I think we should support that as well.

Yes, I omitted it from this PR since we first need to decide how to handle the variables (which GUID to write to). Also, in case we do use Shell Protocol, I am not sure how to go about implementing it since there doesn't seem to be anything providing similar functionality in Shell Protocol. And since we cannot really query what Guid the Shell variables are using, we might need to return every defined variable instead of filtering to a particular GUID.

joboet commented:
Sorry about leaving you hanging, life got in the way.

@dvdhrm and @nicholasbishop, what are your thoughts on this? Should the UEFI target require UEFI-shell, use it if present or not use it at all?

nicholasbishop commented:
I think it really depends on what the intended use case of uefi-std support is. Here are two possibilities:
1. uefi-std is implemented to allow the full rust test suite to run, with the primary goal of catching bugs in the uefi targets.
2. uefi-std is intended to become the normal way you write a UEFI application in Rust, the std support is for end users.

If the goal is 1, then I think targeting the uefi shell sounds good. If the goal is 2, then I don't think it would be very useful to target the uefi shell. The typical UEFI application is a bootloader run directly at boot (or loaded by a previous-stage bootloader), and no shell will be run in that case.

Ayush1325 commented:
> I think it really depends on what the intended use case of uefi-std support is. Here are two possibilities:
> 
> 1. uefi-std is implemented to allow the full rust test suite to run, with the primary goal of catching bugs in the uefi targets.
> 2. uefi-std is intended to become the normal way you write a UEFI application in Rust, the std support is for end users.
> 
> If the goal is 1, then I think targeting the uefi shell sounds good. If the goal is 2, then I don't think it would be very useful to target the uefi shell. The typical UEFI application is a bootloader run directly at boot (or loaded by a previous-stage bootloader), and no shell will be run in that case.

I would like to target 2 while enabling 1. I would like to allow using uefi-std to write general uefi applications (like system test software, maybe uefi shell itself) in Rust. And well, to make 2 viable, we kind of need to be able to run tests, at least on the CI at some point in the future.

So, the current philosophy I have is to implement everything in a way that can be used as widely as possible. That is why as long as something is possible with the base UEFI protocol, I would like to stick with it.

Do note that I am not really sure if there are any companies or any other projects that have an interest in uefi-std (or if they even know of it's existence). I am just doing it in my free time because I did the work to allow running Rust test suit during GSoC 2022 and well, I might as well upstream everything. My time to work on this is quite limited nowadays, but I still have plans to at least upstream File IO, TCP, UDP, and pass args to Command.

joboet reviewed (changes requested):
I needed to do a bit of soul-searching, sorry for that taking so long. I'm inclined to accept this with the current strategy, given
that
* UEFI support is still experimental so we won't be locked in by this
* The implementation is the only reasonable one when UEFI-Shell is unavailable and
* EDK2 publicly documents its GUID, so we can safely interface with it

Please do however implement `Env` using `GetNextVariableName` as well, so that we fully support environment variables.

Ayush1325 commented:
@rustbot ready

joboet commented on library/std/src/sys/pal/uefi/os.rs:
```diff
@@ -171,44 +172,70 @@ pub fn current_exe() -> io::Result<PathBuf> {
-pub struct Env(!);
+#[derive(Clone)]
+pub struct Env {
+    last_var_name: Vec<u16>,
+    last_var_guid: r_efi::efi::Guid,
```
This needs to cache all environment variables, as `Env` is ["[a]n iterator over a snapshot of the environment variables"](https://doc.rust-lang.org/nightly/std/env/struct.Vars.html).

nicholasbishop commented:
> I would like to target 2 while enabling 1. I would like to allow using uefi-std to write general uefi applications (like system test software, maybe uefi shell itself) in Rust. And well, to make 2 viable, we kind of need to be able to run tests, at least on the CI at some point in the future.
> 
> So, the current philosophy I have is to implement everything in a way that can be used as widely as possible. That is why as long as something is possible with the base UEFI protocol, I would like to stick with it.
> 
> Do note that I am not really sure if there are any companies or any other projects that have an interest in uefi-std (or if they even know of it's existence). I am just doing it in my free time because I did the work to allow running Rust test suit during GSoC 2022 and well, I might as well upstream everything. My time to work on this is quite limited nowadays, but I still have plans to at least upstream File IO, TCP, UDP, and pass args to Command.

Thanks for the details, that all makes sense. I really appreciate the work you're doing on this!

Ayush1325 commented:
So, I also did a lot of mulling things over after working on fs implementation and I have started warming upto the inclusion of UEFI Shell Protocol for some things. I am planning to use it in the future for the implementation of `getcwd` and `cwd` since I do not think either make sense when shell is not present.

So, the major questions we need to ask are:
1. Does it make sense to have Environment variables supported when shell is missing?
2. Is it fine to panic in case of a missing shell since `set_var` will panic if it fails?

cc @nicholasbishop @dvdhrm

nicholasbishop commented:
> 1. Does it make sense to have Environment variables supported when shell is missing?

On Zulip, @joshtriplett [wrote](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Environment.20Variables.20for.20UEFI/near/450497021):

> I think it might make sense to have two APIs here.
>
> First, an API under `std::os::uefi::env` that takes a key and GUID, separately.
>
> And second, the standard API in `std::env`, which we'd need to decide on the behavior of; for that, it probably makes sense to do what the shell and other applications do (e.g. what do UEFI ports of other languages like Python do here)?

I think that feels right. The `std::env` public API is appropriate for Unix and Windows, but it just doesn't capture the range of things you can do with UEFI variables. In most cases it's important to have control over a variable's vendor, and whether the variable is volatile, and whether it's accessible during runtime services, and so on.

(Incidentally, I was curious if  [gnu-efi](https://sourceforge.net/projects/gnu-efi/) provides a C-ish API like `getenv`/`putenv`. Seems the answer is no.)

To directly answer "does it make sense to have environment variables supported when shell is missing?": I think requiring the UEFI shell protocol for `std::env` is fine if we also provide Josh's suggested API under `std::os::uefi::env` for full access and control over UEFI variables.

In practice, I think that in the context of UEFI applications the `std::env` API would not often be used, in favor of the more powerful `std::os::uefi::env` API (especially for writing variables, because as described above, in most cases the application does care about the details).

> 2. Is it fine to panic in case of a missing shell since `set_var` will panic if it fails?

Seems OK to me given that the public API provides no other way to return an error, and with the above rational that `std::os::uefi::env` would be more commonly used anyway.

joboet commented:
Nominating for T-libs-api discussion. Do you agree with the changes as described by Nicholas?

Ayush1325 commented:
> Nominating for T-libs-api discussion. Do you agree with the changes as described by Nicholas?

Yes, I will switch to using shell protocol for `std::env` in this PR.

I will create a separate PR for `std::os::env` since it will require adding new public types for UEFI (like GUID, etc), and thus would probably require API related discussion

Ayush1325 committed:
std: uefi: Add basic Env variables

- Implement environment variable functions
- Using EFI Shell protocol.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
@joboet @nicholasbishop I have switched to using shell protocol in this PR.

I will open a seperate PR for variables under `std::os::uefi::env` since that will need discussion regarding the public APIs.

bors commented:
:umbrella: The latest upstream changes (presumably #131747) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Ayush1325 committed:
std: uefi: Use common function for UEFI shell

- Since in almost all cases, there will only be 1 UEFI shell, share the
  shell handle between all functions that require it.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

joboet reviewed (commented):
This is almost ready to go, just two minor things

joboet commented on library/std/src/sys/pal/uefi/os.rs:
```diff
@@ -261,36 +284,66 @@ pub fn getpid() -> u32 {
+    }
+
+    pub(crate) fn get_all_keys() -> io::Result<Vec<OsString>> {
+        let shell = helpers::open_shell().ok_or(unsupported_err())?;
```
Perhaps this should return an empty list in non-shell environments? I'm not sure...

Ayush1325 replied:
I think it is better to return error for consistency because we will be returning error/null in get and set env in this case.

joboet commented on library/std/src/sys/pal/uefi/os.rs:
```diff
@@ -261,36 +284,66 @@ pub fn getpid() -> u32 {
+        if r.is_error() { Err(io::Error::from_raw_os_error(r.as_usize())) } else { Ok(()) }
+    }
+
+    pub(crate) fn get_all_keys() -> io::Result<Vec<OsString>> {
```
I'd fetch the key values in here, so that the key doesn't have to be converted back to UTF-16 in `get`.

Ayush1325 replied:
Nice catch, will do.

joboet reviewed (approved):
I'm fine with these changes as-is and will merge them, but if you want to address the comments here, that's fine with me as well.

joboet commented:
r=me if you want to leave the comments for a separate PR.
@bors delegate+

bors commented:
:v: @Ayush1325, you can now approve this pull request!

If @joboet told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@joboet`
<!-- homu: {"type":"Delegated","delegator":"joboet","delegate":"Ayush1325"} -->

Ayush1325 commented:
@bors r=@joboet

bors commented:
:pushpin: Commit 753536aba86310d85e71bb49062d883938602417 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 753536aba86310d85e71bb49062d883938602417 -->
<!-- homu: {"type":"Approved","sha":"753536aba86310d85e71bb49062d883938602417","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
